department-of-veterans-affairs/caseflow
# Lianna/APPEALS-16339

Resolves [APPEALS-16339](https://vajira.max.gov/browse/APPEALS-16339)

### Description
In order for the frontend to have the information it needs to display the issue categories in its queue columns, non-rating issue category information for each task in the current queue needs serialized to JSON. The following serializers will need to process this data, and have attributes for request issue categories:

[WorkQueue::TaskSerializer](https://github.com/department-of-veterans-affairs/caseflow/blob/master/app/models/serializers/work_queue/task_serializer.rb)
[WorkQueue::TaskColumnSerializer](https://github.com/department-of-veterans-affairs/caseflow/blob/master/app/models/serializers/work_queue/task_column_serializer.rb)
The list of issue categories gathered for each task should only contain at most one instance of a category. For example, let's assume a task is associated with an appeal with the following request issues:

ID: 1 - Category: Camp Lejune Family Member
ID: 2 - Category: Clothing Allowance
ID: 3 - Category: Clothing Allowance
The value of the array of issue categories supplied to the frontend should be:

`{`
	`"issueCategories": [`
		`"Camp Lejune Family Member",`
		`"Clothing Allowance"`
	`]`
`}`
Please coordinate with whomever is assigned [APPEALS-16329](https://vajira.max.gov/browse/APPEALS-16329) around the attribute name you will be using, as they will need it for their column definitions (so they can utilize the correct values for populating their columns).

You will also want to reach out to the developer assigned to [APPEALS-16340](https://vajira.max.gov/browse/APPEALS-16340) to ensure that you both are using the same attribute name.

### Acceptance Criteria
- [ ] An attribute is added to the following serializers for transmitting request issues to the frontend for display in conventional queues:
  * [ ] WorkQueue::TaskSerializer
  * [ ] WorkQueue::TaskColumnSerializer
- [ ] Only unique issue categories are returned from the serializers
- [ ] Add attribute to [taskAttributesFromRawTask](https://github.com/department-of-veterans-affairs/caseflow/blob/master/client/app/queue/utils.js#L74) so that the new values will be placed in tasks objects within our Redux store.
- [ ] Adequate automated test coverage is constructed to validate correct behavior, and guard against regressions.

### Testing Plan
1. Open the Rails Console: `make c`
2. Run the following: `po = VhaProgramOffice.find(43)
task = FactoryBot.create(:assess_documentation_task, assigned_to: po)` 
3. Then run: `task = FactoryBot.create(:vha_document_search_task, assigned_to: po)`
4. Next run: `task.appeal.request_issues << FactoryBot.create(:request_issue, nonrating_issue_category:Constants.ISSUE_CATEGORIES.vha.sample)
task.save`
5. Go to `http://localhost:3000/test/users` 
6. Choose user **VHAPOADMIN**
7. Under the Queue tab Click **Your Queue** 
![16339 Steps 5 to 7](https://user-images.githubusercontent.com/17394371/235500156-dcafee48-898c-4e44-88c7-fff9947b1526.png)
8. On the right above the table click **Switch Views** and select "Prosthetics Cases"
9. With the changes from **APPEALS-16329** there should be a column between **Issues** and **Last Action** titled **Issue Categories**
10. To complete testing go to Terminal and run: `make one-test spec/models/serializers/work_queue/task_serializer_spec.rb:62`
11. Test should pass:
<img width="1671" alt="16339 Testing Suite Pass" src="https://user-images.githubusercontent.com/17394371/235499518-5cfa40f3-c6c0-4c25-ab7a-df119d6fcf21.png">

### User Facing Changes

 BEFORE|AFTER
 ---|---
